Source/Particles/WarpXParticleContainer.cpp

```diff
@@ -572,59 +572,60 @@ WarpXParticleContainer::DepositCharge (WarpXParIter& pti, RealVector const& wp,
                                        const int * const ion_lev,
                                        amrex::MultiFab* rho, int icomp,
                                        const long offset, const long np_to_depose,
-                                       int thread_num, int lev, int depos_lev)
+                                       int thread_num, int lev, int depos_lev,
+                                       bool deposit_on_demand)
 {
-    if (!do_not_deposit) {
-        WarpX& warpx = WarpX::GetInstance();
+    if (do_not_deposit && !deposit_on_demand) return;
```

Comment:
Turn off whitespace in the diff viewer to see this clearly.

Reply:
Looking at this again, it's pretty wild that we patch this here instead on doing the patch at the call side.

Reply:
I choose to implement it this way, and not do that refactoring, because there are a variety of call sites, with somewhat different use cases, and I didn't really want to mess about with them. I think a few of them were outside `WarpXParticleContainer` itself and its descendants, so it felt wrong to have them inspecting the `do_not_deposit` member variable to control their calls to the `DepositCharge` member function.

Reply:
It may be possible to refactor those call sites to use something more like the higher-level `DepositCharge` member that contains the loop over calls to this one, and then apply the flag there.

Reply:
I have to admit I feel the existing `do_not_deposit` inside the call and the addition in this PR are not great design patterns.

There are 10 files with 1-4 calls each that call this function. I really would prefer refactoring and `if`-ing on the PC's `do_not_deposit` for the call.